### ISSUE TITLE
Reason for placing classifier linear hooks

### ISSUE THREAD

**Issue by shashank7289** [JindongGu/Contrastive-LRP#1]
what is the reason that classifier linear forward/backward hooks are only placed on conv2d in layer1 and ReLU from layers 3,6 and not on other conv2d / ReLU layers?

**Comment by JindongGu** [comment]
I am not sure if I understand you correctly.

The VGG16 model consists of two parts (the feature part + the classifier part):

In classifier part, the architecture looks like this
self.classifier = nn.Sequential(
            nn.Linear(512 * 7 * 7, 4096),
            nn.ReLU(True),
            nn.Dropout(),
            nn.Linear(4096, 4096),
            nn.ReLU(True),
            nn.Dropout(),
            nn.Linear(4096, num_classes),
        )
where we only add forward/backward hooks on linear layer. Namely, it corresponds to indice (0, 3, 6).

Simlarly, we add corresponding hooks on conv layer and pooling layer with the help of indices.

There is no need to add a hook on ReLU layer, since the neuron with zero activation receive zero relevance values anyway. More details can be found in our paper and LRP paper.

**Comment by shashank7289** [comment]
thanks a lot.. makes it much clearer.

I am trying to make CLRP work on SSD with Resnet50 as backbone. How can the code be adapted to work with Bottlenecks (i.e. parallel computation branches)?

The code uses a global variable 'R' that stores the current relevance score and is iteratively overwritten during the backward pass. This would not work for parallel computational branches which need to store the relevance values separately.

Do you have a suggestion how could the mathematical formulas be adapted to solve this?
